frontend/src/lib/components/modals/AddProjectUser.svelte

```diff
@@ -0,0 +1,71 @@
+<script lang="ts">
+	import Form from '$lib/forms/Form.svelte';
+	import Input from '$lib/forms/Input.svelte';
+	import t from '$lib/i18n';
+	import Modal, { CloseReason } from './Modal.svelte';
+	import { addProjectUser } from '$lib/services/project.service';
+	import { ProjectRole } from '$lib/gql/graphql';
+
+	export let projectId: string;
+	let modal: Modal;
+	async function openModal(): Promise<void> {
+		if ((await modal.openModal()) === CloseReason.Cancel) return;
```

Comment:
oh ok, I think I get this now...  the `await` here is to force things to be synchronous based upon the user's response... that's what I was having difficulty with here.  It would be helpful to me if we reconsidered the naming here then (or just established a more common approach like offering dispatched events by name, e.g., `on:cancel`.  I thought the await was waiting for it to mount and render and the await was the message that it was open (visible to the user)

Reply:
ahh ok, yeah I could see the thinking that it was awaiting the open to finish. How about renaming it to `openWaitResponse`? or something similar.

I know this isn't the common approach, but I find the common approach lacking in many ways as it splits up your code into a calling function and a receiving function. You open the modal from one place, and you respond to the submit from another place and it makes it really difficult to follow what's happening when you are reading the code because it's split up like that when using events. It does make the open method a little more complicated but I think the tradeoff is worth it. If you look at the Remove example you can really see the benefit because you don't need to pass the delete state from the opening function to the close handler. Though for delete I did handle the API a little different by having a callback that you pass into the open method which gets called on submit.

Reply:
> How about renaming it to `openWaitResponse`? or something similar.

I think that certainly moves things in a clearer direction, you might consider something even more clear like `openAndAwaitUserResponse`... the fact that you are doing multiple things in this method might be an indication of a possible api complexity, e.g., `modal.open() on:response={handleResponse}` might be another option.

> I find the common approach lacking in many ways

It's probably something better to discuss in person but the idea of doing things in two different places seems resolvable by thinking in terms of events bubbling up elements to the consumer who cares, i.e., `createEventHandler`.  From a simplicity standpoint, the consumer expecting to have an event like `submitted`, `cancelled`, `yes`, `no`, etc... keeps the api predictable and following normal event handling and should keep all the handling and logic in the consumer.